File: original-ts-translations/utils-translated.ts
import { Address4 } from 'ip-address';
import { Writable } from 'stream'

export type PacketHeaders = Map<string, opResult>

export enum opResultKind {
    Float,
    Int,
    IPv4,
    MAC,
    Empty,
}

export type opResult = 
    | {kind: opResultKind.Float, val: number}
    | {kind: opResultKind.Int, val: number}
    | {kind: opResultKind.IPv4, val: Address4}
    | {kind: opResultKind.MAC, val: Uint8Array} 
    | {kind: opResultKind.Empty, val: null}

export type Operator = {
    next:  (headers: PacketHeaders) => void;
    reset: (headers: PacketHeaders) => void;
}

export type opCreator    = (nextOp: Operator) => Operator
export type dblOpCreator = (nextOp: Operator) => [Operator, Operator]

export function $π(opCreatorFunc: opCreator, nextOp: Operator): Operator {
    return opCreatorFunc(nextOp);
}

export function $$π(opCreatorFunc: dblOpCreator, nextOp: Operator): [Operator, Operator] {
    return opCreatorFunc(nextOp);
}  

export function tcpFlagsToStrings(flags: number) : string {
    let acc: string = "";
    new Map([
        ["FIN", 1 << 0],
        ["SYN", 1 << 1],
        ["RST", 1 << 2],
        ["PSH", 1 << 3],
        ["ACK", 1 << 4],
        ["URG", 1 << 5],
        ["ECE", 1 << 6],
        ["CWR", 1 << 7],
    ]).forEach((val: number, key: string) => { if ((flags & val) === val) {
                                                    acc = acc + (acc === "") 
                                                    ? `${key}` 
                                                    : "|" + `${key}`;
                                                    }})  
    return acc;
}

export function intOfOpResult(input: opResult) : number | TypeError{
    switch (input.kind) {
        case opResultKind.Int:
            return input.val;
        default:
            throw new TypeError("Trying to extract int from non-int result");
    }
}

export function floatOfOpResult(input: opResult) : number | TypeError {
    switch (input.kind) {
        case opResultKind.Float:
            return input.val;
        default:
            throw new TypeError("Trying to extract float from non-float result");
    }
}

export function stringOfOpResult(input: opResult) : string {
    switch (input.kind) {
        case opResultKind.Float:
        case opResultKind.Int:
        case opResultKind.IPv4:
        case opResultKind.MAC:
            return input.val.toString();
        case opResultKind.Empty:
            return "Empty";
    }
}

export function stringOfPacketHeaders(input_packet: PacketHeaders) : string {
   return ([...input_packet.entries()])
            .reduce((acc, [key, val]) => acc += `${key}" \
                        => ${stringOfOpResult(val)}, `, "");
}

export function packetHeadersOfList(header_list: [string, opResult][]) : PacketHeaders {
    return new Map(header_list);
}

export function dumpPacketHeaders(outc: Writable, headers: PacketHeaders) : void {
    outc.write(`${stringOfPacketHeaders(headers)}`);
}

export function lookupInt(key: string, headers: PacketHeaders) : number | TypeError {
    return intOfOpResult(headers.get(key) ?? 
            {kind: opResultKind.Empty, val: null});
}

export function lookupFloat(key: string, headers: PacketHeaders) : number | TypeError {
    return floatOfOpResult(headers.get(key) ?? 
                {kind: opResultKind.Empty, val: null});
}

File: original-py-translations/utils_translated.py

from typing import Iterator, TextIO, ItemsView, Callable, Union
from abc import ABC
from dataclasses import dataclass
from ipaddress import IPv4Address
from collections import deque
from functools import partial

type op_result_type = Union[float, int, IPv4Address, bytearray, None]

@dataclass
class Op_result(ABC):
    val: op_result_type

    def __hash__(self): 
        return hash(self.val)

@dataclass
class Float(Op_result):
    val: float

@dataclass
class Int(Op_result):
    val: int

    def __hash__(self): 
        return hash(self.val)

@dataclass
class Ipv4(Op_result):
    val: IPv4Address

    def __hash__(self):
        return hash(self.val.exploded)

@dataclass
class MAC(Op_result):
    val: bytearray

    def __hash__(self):
        return hash(bytes(self.val))

@dataclass
class Empty(Op_result):
    val: None = None


class PacketHeaders:
    def __init__(self, data: dict[str, Op_result] | None=None):
        self.data: dict[str, 'Op_result'] | None = data if data is not None else {}

    def __iter__(self):
        return iter(self.data)
    
    def __getitem__(self, key: str):
        if not isinstance(key, str):
            raise TypeError("Packet keys may only be strings.")
        else:
            return self.data[key]
    
    def __setitem__(self, key: str, val: Op_result):
        if not isinstance(key, str):
            raise TypeError("Packet keys may only be strings.")
        elif not isinstance(val, Op_result):
            raise TypeError("Packet values may only be Op_results.")
        else: 
            self.data[key] = val
        return self
    
    def items(self) -> ItemsView[str, Op_result]:
        return self.data.items()
    
    def __hash__(self):
        return hash(frozenset(self.data.items()))
    
    def __or__(self, other: 'PacketHeaders') -> 'PacketHeaders':
        if not isinstance(other, PacketHeaders):
            raise TypeError("Packets may only be unioned with other Packets.")
        return PacketHeaders(self.data.__or__(other.data))
    
    def get(self, key: str, default=None):
        if not isinstance(key, str):
            raise TypeError("Packet keys may only be strings.")
        return self.data.get(key, default)
    
    def get_mapped_int(self, key: str) -> int:
        if not isinstance(key, str):
            raise TypeError("Packet keys may only be strings.")
        return int_of_op_result(self.data[key])
    
    def get_mapped_float(self, key: str) -> float:
        if not isinstance(key, str):
            raise TypeError("Packet keys may only be strings.")
        return float_of_op_result(self.data[key])


class Operator(ABC):
    def __init__(self, next: Callable[[PacketHeaders], None], 
                 reset: Callable[[PacketHeaders], None]):
        self.next = next
        self.reset = reset
    def next(packet: PacketHeaders) -> None:
        raise NotImplementedError("next method not implemented in base class.")
    def reset(packet: PacketHeaders) -> None:
        raise NotImplementedError("reset method not implemented in base class.")

class Op_to_op:
    def __init__(self, func: Callable[[any], Operator], *args):
        self.func = partial(func, *args)
    def __rshift__(self, op: Operator) -> Operator:
        if not isinstance(op, Operator):
            raise TypeError(f"Can only apply an operator \
                                argument to an Op_to_op.")
        return self.func(op)

class Op_to_op_tup(Op_to_op):
    def __init__(self, func: Callable[[any], Operator], *args):
        self.func = partial(func, *args)
    def __rshift__(self, op: Operator) -> tuple[Operator, Operator]:
        if not isinstance(op, Operator):
            raise TypeError(f"Can only apply an operator \
                                argument to an Op_to_op_tup.")
        return self.func(op)

def string_of_mac(buf: bytearray) -> str:
    return f"{buf[0]: .2f}:{buf[1]: .2f}:{buf[3]: .2f}:\
                {buf[4]: .2f}:{buf[5]: .2f}"

def tcp_flags_to_strings(flags: int) -> str:
    tcp_flags: list[str] = [ 
                            "FIN", "SYN", "RST", "PSH", 
                            "ACK", "URG", "ECE", "CWR", 
                           ]
    def tcp_flags_map() -> Iterator[list[tuple[str, float]]]:
        for i, flag in enumerate(tcp_flags):
            yield (flag, 1 << i)
    return "|".join(flag for flag, val in tcp_flags_map() 
                    if (flags & val) == val)

def int_of_op_result(input: Int) -> int:
    match input:
        case Int(val=val): return val
        case _:
            raise TypeError("Trying to extract int from non-int result")

def float_of_op_result(input: Float) -> float:
    match input:
        case Float(val=val): return val
        case _:
            raise TypeError("Trying to extract float from non-float result")
        
def string_of_op_result(input: Op_result) -> str:
    match input:
        case Float(val=val) | Int(val=val) | Ipv4(val=val):
            return f"{val}"
        case MAC(val=val):
            return string_of_mac(val)
        case Empty():
            "Empty"
        case _:
            raise RuntimeError("Reached unreachable code")
        
def string_of_packet(input_packet: PacketHeaders) -> str:
    return "".join(f'"{key}" => {string_of_op_result(val)}, ' 
                   for key, val in input_packet)
        
def packet_of_list(packet_list: deque[tuple[str, Op_result]]) -> PacketHeaders:
    return PacketHeaders({key: val for key, val in list(packet_list)})

def dump_packet(outc: TextIO, packet: PacketHeaders) -> None:
    print(string_of_packet(packet), outc)

def lookup_int(key: str, packet: PacketHeaders) -> int:
    return int_of_op_result(packet[key])

def lookup_float(key: str, packet: PacketHeaders) -> float:
    return float_of_op_result(packet[key])

